chunky/src/java/se/llbit/chunky/chunk/biome/BiomeDataFactory.java

```diff
@@ -0,0 +1,178 @@
+package se.llbit.chunky.chunk.biome;
+
+import se.llbit.chunky.chunk.ChunkData;
+import se.llbit.chunky.world.biome.Biome;
+import se.llbit.chunky.world.biome.BiomePalette;
+import se.llbit.chunky.world.biome.Biomes;
+import se.llbit.log.Log;
+import se.llbit.math.QuickMath;
+import se.llbit.nbt.ListTag;
+import se.llbit.nbt.SpecificTag;
+import se.llbit.nbt.StringTag;
+import se.llbit.nbt.Tag;
+import se.llbit.util.BitBuffer;
+
+import static se.llbit.chunky.world.Chunk.*;
+import static se.llbit.util.NbtUtil.getTagFromNames;
+
+public class BiomeDataFactory {
+  //TODO: Ideally we would have registered factory impls with an isValidFor(Tag chunkTag), but this messy if chain works for now
+  public static void loadBiomeData(ChunkData chunkData, Tag chunkTag, BiomePalette biomePalette, int yMin, int yMax) {
+    BiomeData biomeData = chunkData.getBiomeData();
+    Tag biomeTagsPre21w39a = chunkTag.get(LEVEL_BIOMES);
+    if (!biomeTagsPre21w39a.isError()) { // pre21w39a tag exists
+      if (biomeTagsPre21w39a.isByteArray(X_MAX * Z_MAX)) {
+        if (!(biomeData instanceof SimpleBiomeData2d)) {
+          biomeData = new SimpleBiomeData2d();
```

Comment:
Simple 2d, simple quart 3d, simple 2d, generic3d… None of these are "simple" and we should change the names to something more meaningful. Same for the chunk data class names, but that's another topic.

Maybe just drop `Simple`.

Reply:
yeah I am terrible at naming such classes x)
will do